File: app/src/main/java/com/paradoxo/amadeus/activity/redesign/SplashScreenActivity.java
package com.paradoxo.amadeus.activity.redesign;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.paradoxo.amadeus.R;
import com.paradoxo.amadeus.activity.LoadActivity;
import com.paradoxo.amadeus.activity.MainActivity;
import com.paradoxo.amadeus.dao.BDGateway;
import com.paradoxo.amadeus.util.Arquivo;

import java.io.IOException;

import static com.paradoxo.amadeus.util.Preferencias.appJaFoiAberto;
import static com.paradoxo.amadeus.util.Toasts.meuToast;
import static com.paradoxo.amadeus.util.Util.configurarToolBarBranca;

public class SplashScreenActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash_screen);

        configurarToolBarBranca(this);
        decidirParaOndeVai();


    }

    private void decidirParaOndeVai() {

        if (appJaFoiAberto(this)) {
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    startActivity(new Intent(getApplicationContext(), MainActivity.class));
                    finish();
                }
            }, 5000);
        } else {
           PrimeiroLoad primeiroLoad = new PrimeiroLoad();
           primeiroLoad.execute();

        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out);
    }

    private void cadastrarNomes() {
        Log.e("tag", "Cópia do banco finalizada");

        Intent loadActivity = new Intent(getApplicationContext(), LoadActivity.class);
        startActivity(loadActivity);
        finish();
    }

    @SuppressLint("StaticFieldLeak")
    public class PrimeiroLoad extends AsyncTask<Void, Void, Boolean> {
        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            meuToast(getString(R.string.carregando_banco_inicial),getApplicationContext());
            Log.e("tag", "Iniciando cópia do banco");
            BDGateway.getInstance(getBaseContext());
            // Isso inicializa o banco para que ele possa ser sobrescrito a seguir
        }

        @Override
        protected Boolean doInBackground(Void... voids) {
            try {
                Arquivo.importarBancoPrimeiroUso("Amadeus.db", getApplicationContext());
                Log.e("tag", "Sucesso na cópia do banco");
            } catch (IOException e) {
                Log.e("tag", "Erro ao copiar o banco");
                e.printStackTrace();
                return false;
            }

            return true;
        }

        @Override
        protected void onPostExecute(Boolean bancoCopiado) {
            super.onPostExecute(bancoCopiado);

            if (bancoCopiado) {
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        cadastrarNomes();
                    }
                }, 1000);
            } else {
                meuToast("Falha ao copiar o banco de dados",getApplicationContext());
            }
        }
    }
}

File: app/src/main/java/com/paradoxo/amadeus/activity/SplashScreenActivity.kt
package com.paradoxo.amadeus.activity

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.pm.PackageInfoCompat
import androidx.lifecycle.lifecycleScope
import com.paradoxo.amadeus.R
import com.paradoxo.amadeus.dao.room.AmadeusDatabase
import com.paradoxo.amadeus.dao.room.JsonAssetLoader
import com.paradoxo.amadeus.dao.room.toEntity
import com.paradoxo.amadeus.dao.room.toHistoricoEntity
import com.paradoxo.amadeus.util.Animacoes
import com.paradoxo.amadeus.util.Preferencias
import com.paradoxo.amadeus.util.Preferencias.appJaFoiAberto
import com.paradoxo.amadeus.util.Preferencias.confirmarAberturaApp
import com.paradoxo.amadeus.util.Toasts.meuToast
import com.paradoxo.amadeus.util.Toasts.meuToastLong
import com.paradoxo.amadeus.util.Util.configurarToolBarBranca
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SplashScreenActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash_screen)
        configurarToolBarBranca(this)
        iniciar()
    }

    private fun iniciar() {
        lifecycleScope.launch {
            // ── Etapa 1: dados legados (tabela mensagem) ────────────────────
            val bancoAntigoExiste = withContext(Dispatchers.IO) {
                AmadeusDatabase.getInstance(this@SplashScreenActivity)
                    .mensagemDAO().contarMensagens() > 0
            }

            if (bancoAntigoExiste) {
                startActivity(Intent(this@SplashScreenActivity, RetroCompatibilidadeActivity::class.java))
                finish()
                return@launch
            }

            // ── Etapa 2: carga inicial de assets ───────────────────────────
            verificarBanco(appJaFoiAberto(this@SplashScreenActivity), this@SplashScreenActivity)
        }
    }

    override fun onStart() {
        super.onStart()
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out)
    }

    override fun onStop() {
        super.onStop()
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out)
    }

    companion object {
        private fun trocarTextoPreCarregamentoBanco(context: Activity) {
            context.findViewById<TextView>(R.id.subTituloTextView)?.text =
                context.getString(R.string.carregando_banco)
            Animacoes.animarComFade(context.findViewById(R.id.logoAmadeusImageView), false)
            Animacoes.animarComFade(context.findViewById(R.id.progressoLayout), true)
        }

        private fun atualizarProgresso(porcentagem: Int, context: Activity) {
            context.findViewById<ProgressBar>(R.id.barraProgresso)?.let {
                it.isIndeterminate = false
                it.progress = porcentagem
            }
            context.findViewById<TextView>(R.id.progressoTextView)?.text = "$porcentagem%"
        }

        private fun decidirParaOndevai(context: Activity) {
            if (Preferencias.getPrefString("nomeUsu", context).isNotEmpty()) {
                context.startActivity(Intent(context.applicationContext, MainActivity::class.java))
            } else {
                context.startActivity(Intent(context.applicationContext, ConfigPrimariaActivity::class.java))
            }
            context.finish()
        }

        private fun verificarBanco(bancoJaInserido: Boolean, context: AppCompatActivity) {
            try {
                val pInfo = context.packageManager.getPackageInfo(context.packageName, 0)
                val longVersionCode = PackageInfoCompat.getLongVersionCode(pInfo)

                if (longVersionCode < 18 || !bancoJaInserido) {
                    copiarBanco(context)
                } else {
                    @Suppress("DEPRECATION")
                    Handler().postDelayed({ decidirParaOndevai(context) }, 1000)
                }
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
                meuToast(context.getString(R.string.erro_iniciar_app), context)
            }
        }

        private fun copiarBanco(context: AppCompatActivity) {
            context.lifecycleScope.launch {
                trocarTextoPreCarregamentoBanco(context)
                var sucesso = false

                withContext(Dispatchers.IO) {
                    try {
                        val db = AmadeusDatabase.getInstance(context)
                        val sentencasDao = db.sentencaDAO()

                        val sentencasJson = JsonAssetLoader.carregarSentencas(context)
                        val sentencasHistoricoJson = JsonAssetLoader.carregarSentencasHistorico(context)
                        val entidadesJson = JsonAssetLoader.carregarEntidades(context)

                        val progressoTotal = sentencasJson.size + sentencasHistoricoJson.size + entidadesJson.size
                        var progressoAtual = 0

                        for (sentenca in sentencasJson) {
                            sentencasDao.inserir(sentenca.toEntity())
                            progressoAtual++
                            val pct = ((progressoAtual / progressoTotal.toFloat()) * 100).toInt()
                            withContext(Dispatchers.Main) { atualizarProgresso(pct, context) }
                        }

                        for (sentenca in sentencasHistoricoJson) {
                            sentencasDao.inserirHistorico(sentenca.toHistoricoEntity())
                            progressoAtual++
                            val pct = ((progressoAtual / progressoTotal.toFloat()) * 100).toInt()
                            withContext(Dispatchers.Main) { atualizarProgresso(pct, context) }
                        }

                        for (entidade in entidadesJson) {
                            db.entidadeDAO().inserir(entidade.toEntity())
                            progressoAtual++
                            val pct = ((progressoAtual / progressoTotal.toFloat()) * 100).toInt()
                            withContext(Dispatchers.Main) { atualizarProgresso(pct, context) }
                        }

                        sucesso = true
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }

                if (sucesso) {
                    confirmarAberturaApp(context)
                    verificarBanco(true, context)
                } else {
                    meuToast("Erro ao carregar o banco de dados", context)
                    meuToastLong("Reinicie o app", context)
                }
            }
        }
    }
}
